### PR TITLE
ARROW-15338: [Python] Add `pyarrow.orc.read_table` API

### PR DESCRIPTION
**Pull request by dongjoon-hyun**
This PR aims to add `pyarrow.orc.read_table` like `pyarrow.parquet.read_table`.

https://github.com/apache/arrow/blob/99f7c3cf3e6c2a9555ceff3d48ef73e485ede546/python/pyarrow/parquet.py#L1943

### CODE REVIEW

**Comment by dongjoon-hyun** [comment]
Could you review this please, @sunchao ?

**Comment by dongjoon-hyun** [comment]
cc @williamhyun

**Review by sunchao — APPROVED** [review]
LGTM

**Comment by dongjoon-hyun** [comment]
Thank you so much for your review and approval, @sunchao !

**Comment by dongjoon-hyun** [comment]
Hi, @kszucs . Could you review this, please?

**Comment by kszucs** [comment]
Sure, I'm going to. Also cc @jorisvandenbossche

**Comment by dongjoon-hyun** [comment]
Thank you so much, @kszucs .

File: python/pyarrow/orc.py [hunk 1]
```diff
@@ -175,3 +176,33 @@ def write_table(table, where):
     writer = ORCWriter(where)
     writer.write(table)
     writer.close()
+
+
+def read_table(source, columns=None, filesystem=None):
+    """
+    Read a table from ORC format
```
Comment (jorisvandenbossche) [inline python/pyarrow/orc.py:183]:
```suggestion
    Read a Table from an ORC file.
```

File: python/pyarrow/orc.py [hunk 2]
```diff
@@ -175,3 +176,33 @@ def write_table(table, where):
...
+
+    Parameters
+    ----------
+    source : str, pyarrow.NativeFile, or file-like object
+        If a string passed, can be a single file name or directory name. For
```
Comment (jorisvandenbossche) [inline python/pyarrow/orc.py:188]: For ORC, this implementation of `read_table` does not support reading directories. So the above should only be a single file name.

File: python/pyarrow/orc.py [hunk 3]
```diff
@@ -175,3 +176,33 @@ def write_table(table, where):
...
+        file-like objects, only read a single file. Use pyarrow.BufferReader to
+        read a file contained in a bytes or buffer-like object.
+    columns : list
+        If not None, only these columns will be read from the file. A column
+        name may be a prefix of a nested field, e.g. 'a' will select 'a.b',
+        'a.c', and 'a.d.e'. If empty, no columns will be read. Note
+        that the table will still have the correct num_rows set despite having
+        no columns.
```
Comment (jorisvandenbossche) [inline python/pyarrow/orc.py:196]: This last sentence doesn't actually seem to be true:

```
In [1]: import pyarrow as pa

In [2]: from pyarrow import orc

In [3]: orc.write_table(pa.table({'a': [1, 2, 3]}), "test.orc")

In [4]: result = orc.ORCFile("test.orc").read(columns=[])

In [5]: result.num_rows
Out[5]: 0

In [6]: result.num_columns
Out[6]: 0
```

File: python/pyarrow/orc.py [hunk 4]
```diff
@@ -175,3 +176,33 @@ def write_table(table, where):
...
+    filesystem : FileSystem, default None
+        If nothing passed, paths assumed to be found in the local on-disk
+        filesystem.
```
Comment (jorisvandenbossche) [inline python/pyarrow/orc.py:199]: I know this is copied from the parquet read_table docstring, but I am not sure this is fully correct? I suppose passing a URI for eg a remote file also works?

File: python/pyarrow/tests/test_orc.py [hunk 5]
```diff
@@ -169,6 +171,36 @@ def test_orcfile_empty(datadir):
...
+def test_readwrite(tmpdir):
+    from pyarrow import orc
+    a = pa.array([1, None, 3, None])
+    b = pa.array([None, "Arrow", None, "ORC"])
+    table = pa.table({"int64": a, "utf8": b})
+    file = tmpdir.join("test.orc")
+    orc.write_table(table, file)
+    output_table = orc.read_table(file)
+    assert table.equals(output_table)
+
```
Comment (jorisvandenbossche) [inline python/pyarrow/tests/test_orc.py:183]: Can you add one more read_table call here with eg `columns=["int64"]`, just to make sure that this keyword is properly passed through

**Review by jorisvandenbossche — COMMENTED** [review]
Thanks for the PR! Added some comments

File: python/pyarrow/orc.py (hunk 1 above)
Comment (dongjoon-hyun) [reply python/pyarrow/orc.py:183]: Thanks.

File: python/pyarrow/orc.py (hunk 3 above)
Comment (dongjoon-hyun) [reply python/pyarrow/orc.py:196]: Thank you for pointing out that. Indeed. I didn't notice that.

Comment (dongjoon-hyun) [reply python/pyarrow/orc.py:196]: Since it needs a fix on `ORCFile` itself, can I handle it later in a separate PR, @jorisvandenbossche ?

Comment (dongjoon-hyun) [reply python/pyarrow/orc.py:196]: Oh, it seems that I can fix it in this layer.
```python
In [4]: result = orc.ORCFile("test.orc").read(columns=[])

In [5]: result.num_rows
Out[5]: 0

In [6]: result = orc.ORCFile("test.orc").read(columns=None)

In [7]: result.num_rows
Out[7]: 3

In [8]: result.num_columns
Out[8]: 1
```

File: python/pyarrow/orc.py (hunk 4 above)
Comment (dongjoon-hyun) [reply python/pyarrow/orc.py:199]: It doesn't work like the following.
Screen Shot 2022-01-18 at 8 38 52 PM
.

File: python/pyarrow/tests/test_orc.py (hunk 5 above)
Comment (dongjoon-hyun) [reply python/pyarrow/tests/test_orc.py:183]: Sure!

**Comment by dongjoon-hyun** [comment]
Thank you for review. I addressed your comments. Could you review this once more, @jorisvandenbossche ?

**Comment by dongjoon-hyun** [comment]
Oops. Here is a correction. For the following
- https://github.com/apache/arrow/pull/12153/files#r785785373

Actually, I chose to return `all` columns like `None` for now.

File: python/pyarrow/orc.py (hunk 4 above)
Comment (jorisvandenbossche) [reply python/pyarrow/orc.py:199]: We don't yet support http remote filesystem (urls), but do in general support URIs for implemented filesystems such as S3 (eg `s3://bucket/data.orc` _should_ work I think)

File: python/pyarrow/orc.py (hunk 3 above)
Comment (jorisvandenbossche) [reply python/pyarrow/orc.py:196]: Hmm, it's maybe better to leave it for a separate JIRA, as reading in all columns to then select none of them in a second step might be a bit surprising for users?

**Review by kszucs — APPROVED** [review]
Thanks @dongjoon-hyun!

File: python/pyarrow/orc.py (hunk 3 above)
Comment (dongjoon-hyun) [reply python/pyarrow/orc.py:196]: Thanks, @jorisvandenbossche . Yes, I agree with you that we need to optimize it in both `ORCFile` API and this one together for that case.

File: python/pyarrow/orc.py (hunk 4 above)
Comment (dongjoon-hyun) [reply python/pyarrow/orc.py:199]: Got it.

**Comment by dongjoon-hyun** [comment]
Thank you, @kszucs , @jorisvandenbossche , and @sunchao .

**Pull request closed without merging**